MeshTangents : node to add Tangent & Binormal to a mesh

donboie opened the pull request:
depends on MeshAlgo in Cortex:
https://github.com/ImageEngine/cortex/pull/556

johnhaddon commented on python/GafferSceneTest/MeshTangentsTest.py:
```diff
@@ -0,0 +1,102 @@
+
+		if not os.path.exists( "/tmp/triangle.scc" ) :
+
+			sceneCache = IECore.SceneCache( "/tmp/triangle.scc", IECore.IndexedIO.OpenMode.Write )
```
Actually, I don't think we need any of this SceneCache shenanigans - we can just use an ObjectToScene node. See [PointsTypeTest](https://github.com/GafferHQ/gaffer/blob/master/python/GafferSceneTest/PointsTypeTest.py#L55) for an example.

johnhaddon commented on python/GafferSceneUI/MeshTangentsUI.py:
```diff
@@ -0,0 +1,76 @@
+##########################################################################
+
+import Gaffer
+import GafferUI
```
Unused import.

johnhaddon commented on python/GafferSceneUI/MeshTangentsUI.py:
```diff
@@ -0,0 +1,76 @@
+
+	"description",
+	"""
+	Calculates tangents for a mesh
```
Let's try to make the description a bit more descriptive - ideally all descriptions should provide something that isn't blindingly obvious from the name already. Things it could mention :

- Tangents are computed relative to a UV set
- Tangents are added as primitive variables

johnhaddon commented on python/GafferSceneUI/MeshTangentsUI.py:
```diff
@@ -0,0 +1,76 @@
+		"uvSet" : [
+			"description",
+			"""
+			uvset to used to calculate the tangent vectors.
```
"The UV set used to"?

johnhaddon commented on startup/gui/menus.py:
```diff
@@ -276,6 +276,7 @@ def __shaderNodeCreator( nodeName, shaderName ) :
 nodeMenu.append( "/Scene/OpenGL/Attributes", GafferScene.OpenGLAttributes, searchText = "OpenGLAttributes" )
 nodeMenu.definition().append( "/Scene/OpenGL/Shader", { "subMenu" : GafferSceneUI.OpenGLShaderUI.shaderSubMenu } )
 nodeMenu.append( "/Scene/OpenGL/Render", GafferScene.OpenGLRender, searchText = "OpenGLRender" )
+nodeMenu.append( "/Scene/Object/Mesh Tangents", GafferScene.MeshTangents, searchText = "MeshTangents" )
```
Put this with the other "/Scene/Object" items above so we can see the ordering.

johnhaddon commented on include/GafferScene/MeshTangents.h:
```diff
@@ -0,0 +1,88 @@
+
+	protected :
+
+		virtual bool processesBound() const;
```
Changing the bounding box of an object is an expensive operation, because it means having to propagate the change up through ancestor locations. Since this node doesn't actually affect the bounding box (it just adds primvars, and doesn't modify P), we should ditch these three methods and instead pass the bound through unchanged.

johnhaddon commented on src/GafferScene/MeshTangents.cpp:
```diff
@@ -0,0 +1,155 @@
+	// Fast pass-throughs for things we don't modify
+	outPlug()->attributesPlug()->setInput( inPlug()->attributesPlug() );
+	outPlug()->transformPlug()->setInput( inPlug()->transformPlug() );
+}
```
This is where you would add the bound pass-through. See [MeshType](https://github.com/GafferHQ/gaffer/blob/master/src/GafferScene/MeshType.cpp#L63) for an example.

johnhaddon commented on src/GafferScene/MeshTangents.cpp:
```diff
@@ -0,0 +1,155 @@
+	{
+		outputs.push_back( outPlug()->objectPlug() );
+	}
+	else if( input == outPlug()->objectPlug() )
```
We shouldn't need this block, since we're not going to modify the bound.

johnhaddon commented on src/GafferScene/MeshTangents.cpp:
```diff
@@ -0,0 +1,155 @@
+	std::pair<PrimitiveVariable, PrimitiveVariable> tangentPrimvars = MeshAlgo::calculateTangents( mesh, uvSet, ortho, posPrimVarName);
+	MeshPrimitivePtr meshWithTangents = runTimeCast<MeshPrimitive>( mesh->copy() );
+
+	meshWithTangents->variables["Tangent"] = tangentPrimvars.first;
```
How about exposing these string constants as plugs? If we're providing a uvSet plug, it feels odd to then prevent the user from directing the result to different variables per uvSet. Also, how does everyone feel about the names? "dPdu" and "dPdv" would be the ones I would find most familiar. @andrewkaufman, @danieldresser, @mattigruener, what would your preference be?

johnhaddon replied:
Or are those names bad because they shadow a standard renderer generated primvar in RenderMan?

andrewkaufman replied:
At IE we definitely don't have a proper standard yet for these names, so I could be convinced to call them dPd(uv), but I wonder what artists would think of those names... they're quite technical really (though so is binormal)... @thiasbxl you requested this (indirectly). What do you think would make sense to most people?

donboie replied:
I vote for dPd[s | t] and dPd[uvset_s | uvset_t] when calculated for a uvSet. Perhaps we should remove all primvar renames from this node and require use of rename on the way in & out if required?

johnhaddon replied:
I wish you weren't so fast! I'm not convinced by this automatic naming approach, and I'd appreciate it if we discussed things before you changed Cortex APIs. I feel bad when I ask you to back up, so instead I guess I'm asking you not to charge ahead when there are still questions outstanding...

My reasoning is that having plugs for uTangent/vTangent names makes what the node does far more self-evident. The user can simply scan down the NodeEditor, and see that a specified uvSet will be used to generate a specified pair of primitive variables. This is much more obvious than trying to read the documentation. It's also way more obvious than having to do a ShufflePrimitiveVariables after the fact, after figuring out what it was those automatic names were.

I guess one argument for the automatic generation of names is that it gives us a standard that maybe allows downstream nodes to infer what they're getting, but this doesn't really fit with the general approach of primitive variables just being "along for the ride" (unless they're given specific meaning by the renderer backends). We don't know if the user will deform P after applying the primvars, and we don't even know what purpose they're being added for. So I think it's better to stick with the more lightweight API in Cortex, and give the user explicit control over the naming.

If you disagree, I'd love to hear your reasoning, but prior to code changes!

johnhaddon replied:
Forgot to add - I think `uTangentPlug()` and `vTangentPlug()` with defaults of either "uTangent" and "vTangent" or "sTangent" and "tTangent" would be the most obvious names.

donboie replied:
I suppose I'm using the PR as mechanism to propose speculative code changes and follow the discussion, perhaps this is an abuse of the feature and frustrating.  

What does the phrase "along for the ride" mean in regards to primitive variables?

In retrospect what you describe sounds simple the for  the user to understand, general & cleaner API. I suppose I just went on *tangent* with regards to the derivative notation.

Should I wait for further discussion from other parties before implementing the following:

* add uTangentPlug() & vTangentPlug() which default to uTangent & vTangent. I assume were going to  be consistent using uv rather than st in cortex 10?

andrewkaufman replied:
Those names and defaults sound fine to me. I guess its a bit odd if have uvSetPlug default to "st" though. So maybe we should change the default there to "uv" and have a userDefault at IE set via config?

johnhaddon replied:
> I suppose I'm using the PR as mechanism to propose speculative code changes and follow the discussion, perhaps this is an abuse of the feature and frustrating.

I think it's a pretty decent way of discussing things given we can't do it face to face. It'd just be good to know what's speculative and what's not - maybe a second commit with the speculative bit would have helped isolate it for discussion, and make it easy to roll back or squash it in later.

> What does the phrase "along for the ride" mean in regards to primitive variables?

Sorry, I didn't explain myself very well. I was thinking of two things :

- Once created, a primvar stays as it is - it's just data being passed along. If the mesh was deformed downstream, nothing would magically update the tangents.
- We don't currently assign any meaning to the names of primvars, other than in a couple of spots in the renderer backends (we map s/t to Arnold's uv attribute for instance). Generally (I think I'm right in saying this, but perhaps I should take a look at some of the internal nodes) nodes get told what primvar to operate on, rather than looking for predefined names.

> In retrospect what you describe sounds simple the for the user to understand, general & cleaner API. I suppose I just went on tangent with regards to the derivative notation.

I _think_ it is. I hesitated a bit while writing the above, because in some ways it'd be nice for the downstream nodes to rely on some naming semantics, so they can be like "find the tangents primvars, or compute them if they're not there". But I think that falls down because of the lack of automatic updates when things are deformed - if primvars are just along for the ride, they can't really have well defined semantics. So I think exposing the names to the user on plugs on both the producing and consuming nodes is the best we can do, and as long the default values line up for those, folks won't need to touch them most of the time.

> Should I wait for further discussion from other parties before implementing the following:

>    add uTangentPlug() & vTangentPlug() which default to uTangent & vTangent. I assume were going to be consistent using uv rather than st in cortex 10?

If nobody want to champion automatic semantic naming, then I still think that would be my preference.

If we do have plugs, then I definitely prefer those plug names over the other alternatives - they're reasonably non-technical, and make it clear that the tangents are with respect to the uv set. I'm not 100% sure whether st should be renamed for Cortex 10, but those default values sound fine to me too. If you do make this change, could you update the docs so that we consistently use uTangent/vTangent terminology rather than mix and match with the tangent/bitangent terms?

johnhaddon commented on python/GafferSceneUI/MeshTangentsUI.py:
```diff
@@ -0,0 +1,76 @@
+		"posPrimVarName" : [
+			"description",
+			"""
+			name of the position primVar used to calculate the tangent vectors for.
```
Maybe "The name of the primitive variable for which tangents are computed. For instance, Pref would compute tangents for the reference position (if it exists)." At least something with proper capitalisation...

johnhaddon commented on python/GafferSceneUI/MeshTangentsUI.py:
```diff
@@ -0,0 +1,76 @@
+		"orthogonal" : [
+			"description",
+			"""
+			ensure the tangent and binormal are orthogonal.
```
Maybe "Adjusts the v tangent to be orthogonal to the u tangent." ?

andrewkaufman commented:
> What's the timeframe we need to get this released on?

Its not super urgent, though would be nice if it goes out in the next major version.

> Before we can merge we'll need to tag Cortex and I'll then need to update the gafferDependencies project binaries to include the new version so that the tests will pass on Travis. Either that, or we do some sort of @expectedFailure thing based on hasattr( IECore, MeshAlgo ).

Maybe we should do that so Travis looks happy for now. It'd be nice to get that Cortex tag and gafferDependencies update for the major version release though. Think that's doable?

donboie commented:
Thanks for the feedback & hopefully I addressed most of the review comments.

I've gone with dPds / dPdt for the primitive variable names for now taking into account position name & uvSet name if non standard.

Also I'm not sure if disabling the tests is sufficient as the compile will fail long before we get to the unit test run.

johnhaddon commented:
> Also I'm not sure if disabling the tests is sufficient as the compile will fail long before we get to the unit test run.

Good point. I guess we could #ifdef in MeshTangentsOp.cpp, but maybe we're best off just updating things properly?

donboie commented:
Forgot to answer one of your questions @johnhaddon. I looked at MeshToPoints for my ***inspiration***. This could also have the fast pass through also as the bounds are the same?

johnhaddon commented:
> I looked at MeshToPoints for my inspiration. This could also have the fast pass through also as the bounds are the same?

I thought the same when I first implemented it, but it's actually not the case because the points have a width and type that means the bounds are larger.

johnhaddon commented:
Quick note to say this looks good to me - let's merge once we've got the Cortex side sorted out and I've got the dependencies project updated.

johnhaddon commented:
Looks like this needs a rebase now Don, but once you've pushed that the tests should go through, since Travis is now updated to use the latest Cortex version.
